### ISSUE TITLE
Move functionality from `Peer` class down into `Protocol` class.

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/py-evm#1546]
### What is wrong?

A while back I did a reasonably big effort to get the *EVM* stuff out of the `p2p` module.  One of the harder parts was the `Peer` classes.  This lead us to the idea that the functionality that is currently wrapped up in the `Peer` class should be split.

### How can it be fixed

The basic idea is:

- `Protocol` class gets the low level stuff
- `Peer` class gets the higher level stuff
- `Peer` class gets instantiated with a `Protocol` instance.

So, the `Protocol` class becomes a *Service* and gets instantiated with the reader/writer streams and other connection information.  It also gets the `read_msg` loop and needs to expose an API for consumer to stream those messages.  I believe that a single consumer model is appropriate here.  It seems prudent to have the *base* protocol messages partitioned from the *sub* protocol messages so that we can have a dedicated handler for `Ping/Pong/Disconnect`.

The `Peer` class should be instantiated with a running `Protocol` instance.  Running the peer would *stream* the incoming messages from the `Protocol` instance and handle them however it deems appropriate.

This design **should** simplify our concept of what a `Peer` is, removing the *need* for us to actually go through the handshake/connection by separating/decoupling the networking component.

**Comment by pipermerriam** [comment]
@carver I'm sleep deprived and foggy at the end of my day.  This was originally your idea IIRC so feel free to update this in any way you see fit.

**Comment by pipermerriam** [comment]
See https://github.com/ethereum/py-evm/pull/1300

**Comment by pipermerriam** [comment]
Ah yes!

#1300 reminded me that part of this is to allow supporting multiple sub-protocols which will be necessary when we have the ability to run as an LES server.

**Comment by carver** [comment]
Yeah, this is related, but not *quite* the same as the idea in #1300 - which does a couple things:
1. removes the `peer.context` attribute, which seemed like a messy grab-bag of stuff
2. tries to move *all* sub-protocol-related things off the main peer object, into a `sub_proto` attribute
2. gets us one step closer to being connected to the same peer under multiple sub-protocols
4. break the dependency from `(sub)Protocol -> Peer`
5. the exchange API can continue to get a bit simpler
6. Maybe testing could be a lot easier with types of peers that understand requests

Note that this is all *only* about sub-protocols; it doesn't touch the main p2p protocol.

Original idea writeup: https://github.com/ethereum/py-evm/pull/1294#discussion_r217853545

> If you look at where the properties derived from `context` are used, it's mostly in `ETHProtocol`. That's actually great, and makes a lot of sense: these protocol specific values are only used where we know they should have been set, inside `ETHProtocol` (as opposed to the base peer kind of knowing about them, but not really because it's this opaque type-busting `context` thing).

> So the idea is `ETHProtocol` should actually be the thing that receives these protocol specific connection data on init, as part of the peer factory, which would set curried versions of ETHProtocol on `ETHPeer`s `_supported_sub_protocols`. If you follow that rabbit hole, I think that means you want to break the dependency that protocol has on peer. The subprotocols would expose one unifying API to the peer: `def encode_request(request: BaseRequest) -> (header: bytes, body: bytes)`. Then the peer would use that to expose `peer.send_request(request)`, which uses the protocol encoding under the hood.
